### ISSUE TITLE
Introduce Code Block component

### ISSUE THREAD

**Issue by mcarrano** [patternfly/patternfly#1582]
There is a need to introduce a component for displaying formatted code blocks for easy display. This component will recognize reserved words per the specified language and display color-coded text. 

There was a similar component in PF3 that was based "Google code pretify" https://code.google.com/archive/p/google-code-prettify/wikis/GettingStarted.wiki

**Comment by mcarrano** [comment]
@jknilans Please add any specific project requirements that you have here. @rachael-phillips FYI

**Comment by rachael-phillips** [comment]
@jknilans do you mind also answering these questions below: 

*What product is this impacting?
*Is this currently blocking implementation?
*Is there any product timeline information you can provide?

**Comment by jknilans** [comment]
This new to PF4 component would be used in Red Hat Connect. We had a tight deadline for where this would be used in our interface so we decided to use the [copy-to-clipboard](https://pf4.patternfly.org/components/ClipboardCopy/examples/) component in the short term. Since we're using the copy-to-clipboard component in the short term, we don't have a need by timeline.

Besides the PF3 example provided earlier, the PF4 website uses color-coded code blocks for each component page. It would be nice to see these turned into a component that could be used elsewhere.

**Requirements**

- ability to copy code block
- color-coded text
- ability to display human-readable JSON text (lines of code don't wrap)
- expand/collapse (nice to have)

**Comment by ncameronbritt** [comment]
This may be a separate issue, but I would also like the ability to style bits of code inline, for example the way `certain words` can be styled as code `inline`. Versus:
```
# Blocks of specially formatted code
Name  Date  Thing
Joe   4/3   Rock
```          

In markdown it's the difference of using one tick versus three.

**Comment by mcarrano** [comment]
@tlabaj @mcoker interested in your thoughts on this one. Do you think this is something we would design and build fro scratch or would we more likely support a third-party solution? Seems like parsing a block of text and color coding it is complex and becomes language dependent, but perhaps there's a simple solution?

**Comment by mcoker** [comment]
We can introduce something more simple in core that allows for both inline and block level formatting of code blocks that does things like use a monospace font, preserves white space, etc. However if we want to enable syntax highlighting, that seems like something we would definitely want to use some sort of library for.

**Comment by mcarrano** [comment]
Per comment by @jknilans above, could we just use the same formatter we use for displaying code examples in the workspace?

**Comment by redallen** [comment]
@mcarrano That formatter is [PrismJS](https://prismjs.com/), which is JS based. Syntax highlighting is complicated and without some verbose markup I'm not sure it's possible to do entirely in HTML/CSS.

This is how Prism html looks:
```html
<pre class="prism">
  <div class="token-line">
    <span class="token tag punctuation">&lt;</span>
    <span class="token tag">nav</span>
    <span class="token tag"> </span>
    <span class="token tag attr-name">class</span>
    <span class="token tag attr-value punctuation">=</span>
    <span class="token tag attr-value punctuation">"</span>
    <span class="token tag attr-value">pf-c-nav pf-m-dark</span><span class="token tag attr-value punctuation">"</span>
    <span class="token tag"> </span>
    <span class="token tag attr-name">aria-label</span>
    <span class="token tag attr-value punctuation">=</span>
    <span class="token tag attr-value punctuation">"</span>
    <span class="token tag attr-value">Global</span>
    <span class="token tag attr-value punctuation">"</span>
    <span class="token tag punctuation">&gt;</span>
    <span class="token plain"></span>
  </div>
</pre>
```

That's just for a single `token-line`:
![image](https://user-images.githubusercontent.com/47335686/73679648-d28a5880-4688-11ea-9766-7d98bad84c19.png)

Maybe we can implement the component first in React with syntax highlighting for a few languages? We could even create a PatternFly style config for an existing highlighter like [HighlightJS](https://highlightjs.org/) , [hilite.me](https://github.com/alexkay/hilite.me), or [PrismJS](https://prismjs.com/) and wrap it as an extension package.

**Comment by mcarrano** [comment]
Interesting approach @redallen . I could see us starting this in React. I will defer to @LHinson about priority. Not sure where this currently lives on our roadmap.

**Comment by gdoyle1** [comment]
Discussed this in the web design meeting today and came up with a design for a code block/snippet that can be copied partly or wholly. Here's the marvel project: https://marvelapp.com/547783a and the sketch file: https://sketch.cloud/s/R3gqz

I'll be making some iterations to the design with a version that uses a text label with copy to clipboard button, or button on left, or options for copy all / copy selected and somehow figuring out a way to approach copying pieces vs a whole

**Comment by mindreeper2420** [comment]
To add on to what Gina said, we've started to develop a version of this for the developer site, using PrismJS with custom styling and JS updates. In choosing PrismJS, we did take into account whether or not it can be used in Drupal (our CMS), React, and it's overall footprint. In addition to base PrismJS, the Line Numbers, Toolbar and Copy to Clipboard plugins are used. The language identifier in Gina's designs is pulled from the `data-lang=""` attribute.

For styling, we are using a select set of token classes in order to both customize the code snippets, but also keep them manageable (as mentioned above). Right now, we support **Markup / HTML / SVG / CSS / JavaScript / Git / AsciiDoc / Bash + Shell / JSON / SCSS / Liquid / Markdown / TOML / Twig / YAML** for syntax highlighting.
